### PR TITLE
Use sandboxes to load project and test dependencies

### LINKED ISSUE #9: test() doesn't load dependencies
If generating a report for a Package that has dependencies that are not in the current environment, TestReports.test() will fail.

To reproduce, dev Example.jl and add a new dependency in `Project.toml`, for example:
```toml
[extras]
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
Dictionaries = "85a47980-9c8c-11e8-2b9f-f7ca1fa99fb4"

[targets]
test = ["Test", "Dictionaries"]
```

And make sure it is used in `runtests.jl`. Output is:

```Julia
┌ Warning: ERROR: Test(s) failed or had an error in Example
└ @ TestReports ~/repos/TestReports.jl/src/runner.jl:222
ERROR: Example had test errors
```

`Pkg.test` handles this by using a sandbox, but we have the added complication of needing to add TestReports to use the ReportingTestSet. I've tested with 1.4 and adding this before the runner code works, like so:

```Julia
            Operations.sandbox(ctx, pkgspec, source_path, test_folder, test_project_override) do
            # test_fn !== nothing && test_fn()
            Pkg.Display.status(Context(), mode=PKGMODE_MANIFEST)
            flush(stdout)
            try
                Pkg.add("TestReports")
                cmd = ```
                    $(Base.julia_cmd())
                    --code-coverage=$(coverage ? "user" : "none")
                    --color=$(Base.have_color ? "yes" : "no")
                    --compiled-modules=$(Bool(Base.JLOptions().use_compiled_modules) ? "yes" : "no")
                    --check-bounds=yes
                    --depwarn=$(Base.JLOptions().depwarn == 2 ? "error" : "yes")
                    --inline=$(Bool(Base.JLOptions().can_inline) ? "yes" : "no")
                    --startup-file=$(Base.JLOptions().startupfile != 2 ? "yes" : "no")
                    --track-allocation=$(("none", "user", "all")[Base.JLOptions().malloc_log + 1])
                    $(runner_file_path)
                    ```
                run(cmd)
                @info "$pkg tests passed. Results saved to $logfilename."
            catch err
                push!(pkgs_errored, pkg.name)
            end
        end
```

I haven't looked to see how 1.0.5 deals with this but guessing its similar. I'll try to do this over the next couple of days and get a PR up, any suggestions are welcome.

### PR DESCRIPTION
**Pull request by mmiller-max**
PR implements sandbox approach from Pkg.test() which runs all tests in temporary environments, installing package and test dependencies as required.

Closes #9

### CODE REVIEW

**Comment by mmiller-max** [comment]
Have requested JuliaTesting access to AppVeyor so we can check the Windows builds.

**Comment by johnnychen94** [comment]
> Have requested JuliaTesting access to AppVeyor so we can check the Windows builds.

Done.

FYI, people are gradually switching from appveyor to Github Actions as it significantly reduces CI waiting time with more concurrent jobs.

**Comment by mmiller-max** [comment]
Thanks for that. Will look at moving it to Github Actions, sounds promising.

**Comment by mmiller-max** [comment]
@oxinabox in particular I'd like to check what you think of the console output with this PR. It makes it a lot more untidy, as TestReports have to be added to the sandbox environment (wouldn't be an issue if `ReportingTestSet` was in Base)

**Comment by oxinabox** [comment]
> as TestReports have to be added to the sandbox environment (wouldn't be an issue if ReportingTestSet was in Base)

I think what we can do is used a stacked enviroment.
Since we are already in an enviroment that is TestReports, we can do:
```julia
TestReportsProjectDir = dirname(@__DIR__))
pushfirst!(Base.LOAD_PATH)
```

before we  do  everything to do with activating the Testing Enviroment.

And thus would not need to install a version of TestReports.
(and thus would not have issues with what version we get)

**Comment by mmiller-max** [comment]
> I think what we can do is used a stacked environment.

Yes that worked well, good idea. I had to add it to the runner code as the sandboxes overwrite `LOAD_PATH` as they create the test environment. I rebased onto master, but you can see the changes in the second commit.

I need to work out why some of the tests are failing. Seems dependencies aren't being loaded as I expected them to be in the test environments.

**Comment by mmiller-max** [comment]
OK think this is there now. When you're happy @oxinabox, I'll squash to one commit.

I was wondering if you wanted to [enforce a linear commit history](https://docs.github.com/en/github/administering-a-repository/requiring-a-linear-commit-history) with this repository? I think it's cleaner, and we could also require the commit message to have the issue number, similar to the main Julia repository?

**Comment by oxinabox** [comment]
I like a linear history, I have never seen it enforced though.
I actually question how valuable it really is, but i do always stick to it in my own PRs.
(I suspect that many conflicts are easier to resolve if just dealing with merges rather than rebases.)

File: src/runner.jl [hunk 1]
```diff
@@ -167,28 +175,66 @@ function test!(pkg::AbstractString,
...
+            --compiled-modules=$(Bool(Base.JLOptions().use_compiled_modules) ? "yes" : "no")
+            --check-bounds=yes
+            --depwarn=$(Base.JLOptions().depwarn == 2 ? "error" : "yes")
+            --inline=$(Bool(Base.JLOptions().can_inline) ? "yes" : "no")
+            --startup-file=$(Base.JLOptions().startupfile != 2 ? "yes" : "no")
+            --track-allocation=$(("none", "user", "all")[Base.JLOptions().malloc_log + 1])
+            --eval $(runner_code)
+            ```
+
+        test_folder_has_project_file = isfile(replace(testfilepath, "runtests.jl"=>"Project.toml"))
```
Comment (oxinabox) [inline src/runner.jl:194]: Maybe tp be sure it is not a very weird path or package name:
```suggestion
        test_folder_has_project_file = isfile(replace(testfilepath, r"/runtests.jl$"=>"/Project.toml"))
```

File: src/runner.jl [hunk 2]
```diff
@@ -167,28 +175,66 @@ function test!(pkg::AbstractString,
...
+                                joinpath(pkgspec.path, "test"),
+                                test_project_override)
+            else
+                sandbox_args = (ctx,
+                                pkgspec,
+                                pkgspec.path,
+                                joinpath(pkgspec.path, "test"))
+            end
+
+            sandbox(sandbox_args...) do
```
Comment (oxinabox) [inline src/runner.jl:214]: shorter and while technically slower, negligably so.
```suggestion
            sandbox_args = (ctx,
                            pkgspec,
                            pkgspec.path,
                            joinpath(pkgspec.path, "test"))   
            if VERSION >= v"1.4.0" && !test_folder_has_project_file
                test_project_override = gen_target_project(ctx, pkgspec, pkgspec.path, "test")
                sandbox_args = (sandbox_args..., test_project_override)
            end

            sandbox(sandbox_args...) do
```

File: src/runner.jl [hunk 3]
```diff
@@ -167,28 +175,66 @@ function test!(pkg::AbstractString,
...
+                # test_fn !== nothing && test_fn()  # TODO: implement
+                flush(stdout)
```
Comment (oxinabox) [inline src/runner.jl:216]: Is there a reason this is needed?
Normally it self-flushes just fine.

File: test/utils.jl [hunk 4]
```diff
@@ -0,0 +1,134 @@
+using Pkg, Test, LibGit2
+
+"""
+`copy_test_package` copied from Pkg.jl as cannot be imported during tests.
```
Comment (oxinabox) [inline test/utils.jl:4]: why not?

File: src/runner.jl (hunk 3 above)
Comment (mmiller-max) [reply src/runner.jl:216]: Not sure, I was copying it directly from [this line](https://github.com/JuliaLang/Pkg.jl/blob/master/src/Operations.jl#L1542) from `Pkg.jl/src/Operations.jl`:

Can definitely remove if you think its not needed?

File: src/runner.jl (hunk 2 above)
Comment (mmiller-max) [reply src/runner.jl:214]: Yeah that's cleaner. The inner `if` condition still needs to be

```Julia
if VERSION >= v"1.4.0" 
    test_project_override = test_folder_has_project_file ?
        nothing :
        gen_target_project(ctx, pkgspec, pkgspec.path, "test")
    sandbox_args = (sandbox_args..., test_project_override)
end
```
as `sandbox` needs the fifth argument in V1.4, so I've merged your suggestion and that.

File: test/utils.jl (hunk 4 above)
Comment (mmiller-max) [reply test/utils.jl:4]: It is defined in `Pkg.jl/test/utils.jl` so to load it we'd need to `include` it (its not available as in import from `Pkg`). However I can't get the package directory of `Pkg` as `pkgdir(Pkg)` returns `nothing`. Unless there is another way of getting the path I don't think we can use it.

**Comment by mmiller-max** [comment]
> I like a linear history, I have never seen it outright enforced though.
> I actually question how valuable it really is, but i do always stick to it in my own PRs.
> (I suspect that many conflicts are easier to resolve if just dealing with merges rather than rebases.)

For looking through the history I think it helps, however the issue tracking in Github probably helps with that greatly too. I'll use it on my PRs too.

> Doesn't seem required.
> GitHub now fairly reliably links commits back to PRs now,
> and PRs link back to issues.

Great!

File: test/utils.jl (hunk 4 above)
Comment (oxinabox) [reply test/utils.jl:4]: Ah it is in the test dir

```suggestion
`copy_test_package` copied from [`Pkg.jl/test/utils.jl`](https://github.com/JuliaLang/Pkg.jl/blob/v1.4.2/test/utils.jl#L209).
```

**Pull request merged**